### PR TITLE
Add getTDDList method to DesmoHub module

### LINKED ISSUE #13: Add enumerate TDDs function
We recently added new functionality in https://github.com/vaimee/desmo-contracts to list all the TDD registered. We should port this feature also in the SDK.

### PR DESCRIPTION
**Pull request by iosonopersia**
Closes #13.

### CODE REVIEW

File: src/lib/desmoHub-module.ts [hunk 1]
```diff
@@ -322,4 +322,38 @@ export class DesmoHub {
     const { url, owner, disabled, score }: ITDD = await this.contract.getTDD();
     return { url, owner, disabled, score } as ITDD;
   }
+
+  /**
+   * @param start the initial TDD index to retrieve (by default is 0)
+   * @param end the final TDD index to retrieve (by default is the index of the last TDD in contract's storage)
+   * @returns an array containing the requested TDD descriptions.
+   */
+  public async getTDDList(start?: number, end?: number): Promise<ITDD[]> {
```
Comment (relu91) [inline src/lib/desmoHub-module.ts:331]: Typescript has the default syntax for function arguments:
```suggestion
  public async getTDDList(start: number=0, end?: number): Promise<ITDD[]> {
```
In a sense is less safe than the `??` that you are using below but it clearly states the default. I would suggest using that here *and* use the  `??`  if somebody is so nasty that it passes `null` as `start` input.

File: src/lib/desmoHub-module.ts [hunk 2]
```diff
@@ -322,4 +322,38 @@ export class DesmoHub {
...
+    const promises: Promise<ITDD>[] = [];
+    for (let i = start; i < end + 1; ++i) {
+      promises.push(this.contract.getTDDByIndex(i));
+    }
+    const tddList = await Promise.all(promises);
+
+    for (let i = 0; i < tddList.length; ++i) {
+      const { url, owner, disabled, score }: ITDD = tddList[i];
+      tddList[i] = { url, owner, disabled, score } as ITDD;
+    }
```
Comment (relu91) [inline src/lib/desmoHub-module.ts:355]: Why are u simply casting each element of the list to ITDD?

File: src/test/desmoHub-test.ts [hunk 3]
```diff
@@ -48,6 +48,13 @@ describe('DesmoHub Tests', function () {
     });
   });
 
+  describe('Retrieve TDD list', function () {
+    it('should retrieve a list of tdds', async () => {
+      const tddList = await desmohub.getTDDList(1, 3);
+      expect(tddList.length).to.equal(3);
+    });
+  });
+
```
Comment (relu91) [inline src/test/desmoHub-test.ts:57]: Need also failing tests for an empty response, zero length, start greater than the end. Start greater than TDD length.

**Review by relu91 — CHANGES_REQUESTED** [review]

File: src/lib/desmoHub-module.ts (hunk 2 above)
Comment (iosonopersia) [reply src/lib/desmoHub-module.ts:355]: The objects that we get from the contract function are of a strange type.
The `url`, `owner`, `disabled` and `score` values are accessible both using a numerical index and using the property name.
For example, the `owner` can be obtained both as `tddObject[1]` and as `tddObject.owner`.

Example:
```typescript
[
    'https://desmold-zion-2.vaimee.it',
    '0x70997970C51812dc3A010C7d01b50e0d17dc79C8',
    false,
    BigNumber { _hex: '0x00', _isBigNumber: true },
    url: 'https://desmold-zion-2.vaimee.it',
    owner: '0x70997970C51812dc3A010C7d01b50e0d17dc79C8',
    disabled: false,
    score: BigNumber { _hex: '0x00', _isBigNumber: true }
  ]
```

What I'm doing in those lines of code is to clean up those objects from the numeric properties.
**I'm doing the same in the `getTDD()` method.**

File: src/test/desmoHub-test.ts (hunk 3 above)
Comment (iosonopersia) [reply src/test/desmoHub-test.ts:57]: I cannot test the "empty TDDStorage" case as it would require to delete the registered TDDs in the deployed contract, which is not possible since the contract doesn't allow for TDD deletions.

I will implement all the other suggested cases.

File: src/lib/desmoHub-module.ts [hunk 4]
```diff
@@ -1,325 +1,374 @@
...
+    const tddList = await Promise.all(promises);
+
+    // Clean up the TDD objects from their numeric properties:
+    for (let i = 0; i < tddList.length; ++i) {
+      const { url, owner, disabled, score }: ITDD = tddList[i];
+      tddList[i] = { url, owner, disabled, score } as ITDD;
+    }
+
+    return tddList;
```
Comment (relu91) [inline src/lib/desmoHub-module.ts:372]:
```suggestion
    const tddList = await Promise.all(promises);

   return tddList.map( ({ url, owner, disabled, score }) => { url, owner, disabled, score } ) 
```
Ok, now I see why are you cleaning up the list. However, using the map method is better than the for a loop. See the example above ( I wrote it by heart, check if it works)

File: src/lib/desmoHub-module.ts [hunk 5]
```diff
@@ -1,325 +1,374 @@
...
+    start = Math.max(start, 0);
+    stop = Math.min(stop, tddStorageLength);
+
+    if (start >= tddStorageLength) {
+      throw new Error(`Start index must be lower than the TDD storage length (${tddStorageLength}).`);
+    } else if (start > stop) {
+      throw new Error(
+        `Start index (${start}) is greater than stop index (${stop}).`,
+      );
+    }
```
Comment (relu91) [inline src/lib/desmoHub-module.ts:352]: No need to use `else if` here, do a regular `if` is cleaner.

File: src/lib/desmoHub-module.ts [hunk 6]
```diff
@@ -1,325 +1,374 @@
...
+     */
+
+    const promises: Promise<ITDD>[] = [];
+    for (let i = start; i < stop; ++i) {
+      promises.push(this.contract.getTDDByIndex(i));
+    }
+    const tddList = await Promise.all(promises);
+
+    // Clean up the TDD objects from their numeric properties:
+    for (let i = 0; i < tddList.length; ++i) {
```
Comment (relu91) [inline src/lib/desmoHub-module.ts:367]:
```suggestion
    for (let i = 0; i < tddList.length; i++) {
```
a more mainstream for-loop.

**Review by relu91 — CHANGES_REQUESTED** [review]
Looks better, but be careful when you manipulate git commits desmoHub-module now it is completely replaced. Is not easy to review and follow the history of changes.

File: src/test/desmoHub-test.ts (hunk 3 above)
Comment (relu91) [reply src/test/desmoHub-test.ts]: Ok, then it is an issue of SDK tests. The unit tests should not rely on the deployed contract but they should use a mock-up. Perhaps we should introduce hardhat as a dev dependency and write SDK tests with that.

File: src/lib/desmoHub-module.ts (hunk 6 above)
Comment (iosonopersia) [reply src/lib/desmoHub-module.ts]: Not needed because we chose to replace the for loop with a call to `map()`.

File: src/lib/desmoHub-module.ts (hunk 5 above)
Comment (iosonopersia) [reply src/lib/desmoHub-module.ts]: Done.

File: src/lib/desmoHub-module.ts (hunk 4 above)
Comment (iosonopersia) [reply src/lib/desmoHub-module.ts]: Done.

**Review by relu91 — APPROVED** [review]
Let's go
